src/shared/generator.h

```diff
@@ -49,3 +51,23 @@ int generator_hook_up_growfs(
         const char *target);
 
 void log_setup_generator(void);
+
+static inline void _assign_arg(const char **where, char *what) {
+        if (where != NULL)
+                *where = what;
+}
+
+/* Similar to DEFINE_MAIN_FUNCTION, but initializes logging and assigns positional arguments. */
+#define DEFINE_MAIN_GENERATOR_FUNCTION(impl, dest_main, dest_early, dest_late) \
+        _DEFINE_MAIN_FUNCTION(impl, r < 0 ? EXIT_FAILURE : EXIT_SUCCESS, ({ \
+                log_setup_generator();                                  \
+                if (argc > 1 && argc != 4)                              \
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), \
+                                               "This program takes zero or three arguments."); \
+                if (argc > 1)                                           \
+                        _assign_arg((dest_main), argv[1]);              \
+                if (argc > 2)                                           \
+                        _assign_arg((dest_early), argv[2]);             \
+                if (argc > 3)                                           \
+                        _assign_arg((dest_late), argv[3]);              \
+                }) )
```

Comment:
hmm, wouldn't it be nicer if instead we'd call the specified main() function with three args for main/early/late? i think this would be a lot more natural.

Reply:
I didn't think about this. I'll try to rework like this.

Reply:
Actually, this doesn't work, because the `arg_dest*` args are static arguments. If they were passed to `run`, it'd have to assign the global args. This is much more code, and makes the patch less nice.

Reply:
hmm? not following. my assumption would be that generators would define a main function like this:

```
int run(const char *main_dir, const char *early_dir, const char *late_dir) {
        arg_dest = main_dir;
        …
};
```

That sounds pretty OK to me, no?

Reply:
It'd have to be:
```C
   if (main_dir)
        arg_dest = main_dir;
```
I *can* rework it like this, but I don't think that's prettier.

Reply:
well. arg_dest generally defaults to /tmp for our generators... this could be handled by the macro too